src/librustc_resolve/lib.rs

```diff
@@ -788,6 +792,59 @@ enum ParentLink<'a> {
     BlockParentLink(Module<'a>, NodeId),
 }
 
+#[derive(Clone, Copy)]
+/// Records information about the resolution of a name in a module.
+struct NameResolution<'a> {
+    /// The number of unresolved single imports that could define the name.
+    outstanding_references: usize,
+    /// The least shadowable known binding for this name, or None if there are no known bindings.
+    binding: Option<&'a NameBinding<'a>>,
+}
+
+impl<'a> Default for NameResolution<'a> {
+    fn default() -> Self {
+        NameResolution { outstanding_references: 0, binding: None }
+    }
+}
+
+impl<'a> NameResolution<'a> {
+    fn result(&self, outstanding_globs: usize) -> ResolveResult<&'a NameBinding<'a>> {
+        // If no unresolved imports (single or glob) can define the name, self.binding is final.
+        if self.outstanding_references == 0 && outstanding_globs == 0 {
+            return self.binding.map(Success).unwrap_or(Failed(None));
+        }
+
+        if let Some(binding) = self.binding {
+            // Single imports will never be shadowable by other single or glob imports.
+            if !binding.defined_with(DefModifiers::GLOB_IMPORTED) { return Success(binding); }
+            // Non-SHADOWABLE glob imports will never be shadowable by other glob imports.
+            if self.outstanding_references == 0 && !binding.defined_with(DefModifiers::SHADOWABLE) {
+                return Success(binding);
+            }
+        }
+
+        Indeterminate
+    }
+
+    // Define the name or return the existing binding if there is a collision.
+    fn try_define(&mut self, binding: &'a NameBinding<'a>) -> Result<(), &'a NameBinding<'a>> {
+        let old_binding = match self.binding {
+            Some(_) if binding.defined_with(DefModifiers::SHADOWABLE) => return Ok(()),
```

Comment:
>Perhaps I should rename SHADOWABLE to PRELUDE.

That would be much clearer, because prelude and shadowable glob imports would behave differently (I suppose).
The resolution result depending on the resolution order is still pretty bad, even if the concern is theoretical at the moment due to absence of user-defined preludes.